### ISSUE TITLE
Proposal: Autolink with `[code]`

### ISSUE THREAD

**Issue by wojtekmach** [elixir-lang/ex_doc#1663]
Currently we autolink expressions like `` `String` ``, `` `String.upcase/1` ``, etc. The obvious problem is auto linking "Erlang" modules, e.g.: `` `:file` ``, which we don't do, because we don't know if the user intended to link to a module or just wrote the atom `:file`. We can today solve this by writing `` [`:file`](`:file`) `` but it's verbose as well as when any other Markdown parser sees it, it will output garbage.

Instead of implicitly treat `` `code` `` as potential links, I'd like to propose supporting an explicit syntax: `` [`code`] ``. This is what [Rust does](https://doc.rust-lang.org/rustdoc/write-documentation/linking-to-items-by-name.html). See https://github.com/erlef/build-and-packaging-wg/issues/60#issuecomment-1347292105. By having explicit syntax, the tooling will know the user intended to write a code reference and thus can output warnings, etc. This will have much more predictable result on non-ExDoc: GitHub, etc would render `` [`:file`] `` as `[<code>:file</code>]` which I think is very reasonable.

Here's a proof of concept: https://gist.github.com/ef18199a4283d9a59b9fdfb571a4e251.

Per https://github.com/erlef/build-and-packaging-wg/issues/60, there are conversations about introducing another EEP that would make Markdown officially supported document format for Erlang. I think there's appetite for supporting the syntax I'm proposing there. If it would be accepted there then I think there'd be a strong reason to support it in Elixir ecosystem too.

**Comment by josevalim** [comment]
I am fine with supporting this but only after we agree on the format for Erlang so we don't end up with three distinct ways in the future. :)

**Comment by Eiji7** [comment]
Interesting! How about change it by using `sigil`-like syntax and modifiers?

For example:
1. default `module` or `mfa` (without optional modifier): <code>\~\`String\`</code> or <code>\~\`String.upcase/1\`</code> or <code>\~\`:file\`</code>
2. same `module` or `mfa` (with `m` modifier): <code>\~\`String\`m</code> or <code>\~\`String.upcase/1\`m</code> or <code>\~\`:file\`m</code>
3. Elixir `forum` topic: <code>\~\`53184\`f</code> or <code>\~\`elixir-v1-14-3-released\`f</code><br/>**Note**: both `elixirforum.com/t/:id` and `elixirforum.com/t/:slug` redirects to `elixirforum.com/t/:slug/:id`
4. `githab` links: <code>\~\`elixir-lang/ex_doc#1663\`g</code> or <code>\~\`elixir-lang/ex_doc@4b5c414\`g</code> or <code>\~\`elixir-lang/ex_doc@tag-v0.29.3\`g</code>
5. `hex` package: <code>\~\`phoenix\`h</code>
6. `hex` documentation: <code>\~\`phoenix\`d</code> or  <code>\~\`phoenix-1.7.0\`d</code>
7. `wikipedia` article: <code>\~\`Elixir_(programming_language)\`w</code> or <code>\~\`pl.Elixir_(język_programowania)\`w</code>

and so on …

**Comment by wojtekmach** [comment]
These are interesting ideas but remember that one of the goals is whatever we write in the docs should be understandable when rendered by Markdown readers other than ExDoc: IEx, GitHub, etc. In those cases, `~`Elixir_(programming_language)`w` is too opaque.

**Comment by Eiji7** [comment]
I see … nobody said that `modifier` needs to be a single letter. The rest is about `~` before said link. For me it just look better, but we can do similar for square brackets as well, right?

For example: <code>[wiki:\`Elixir_(język_programowania)\`:pl]</code>. Maybe even `:pl` is looks better here, so it's more clear that's an option for `wiki` modifier and also we do not need to parse it from the beginning of wiki `slug`.

What do you think?

**Comment by wojtekmach** [comment]
Sorry, I'm not sold writing ``` [wiki:`Elixir_(język_programowania)`:pl] ``` is better than ``` https://pl.wikipedia.org/wiki/Elixir_(język_programowania) ```. I'd personally choose the latter every time as it's unambiguous and interoperable.

**Comment by hauleth** [comment]
> Wild @hauleth appeared

Maybe, just maybe, we could do something to extend the linking support in general by using custom URIs in classic `[]()` and/or `<>` MD constructs. So we would support configuration option to the ExDoc generator like `:link_schemas` that would accept keyword list/map in form of:

```elixir
%{
  "erlang" => ExDocsErlang.URIHandler
}
```

`ExDocsErlang.URIHandler` would then would look like:

```elixir
def module ExDocsErlang.URIHandler do
  # This is used in case of `<erlang:crypto>` to describe what should be show,
  # defaults to just displaying `foo` or `erlang:crypto`, that is up to the discussion.
  def display_name(value) do
    # Here `value` is `"crypto"`, aka the part of the "URI" with "schema" trimmed
    ":#{value}"
  end

  def url(value) do
    # Here we construct `%URI{}` or just string that will link to the given page
  end
end
```

By default ExDoc would provide few "built-in" handlers like:

- `http` and `https` (either as handlers or as just "magic") - regular HTTP links
- `xref` - cross page references within the same generation unit
- `elixir` - link to the Elixir's documentation

So instead of introducing some new magic syntax we will just extend existing one. So current `` `String` `` could be written as `<xref:String>` as well (assuming that we generate documentation for Elixir or `<elixir:String>` if we generate documentation for other packages and we want to link to official Elixir's documentation. In addition to that users could define their own schemas, which would allow additional features to be added in the future, like for example I see potential for:

- `github`/`gitlab` - link to issues and PRs/MRs from the documentation
- `livebook` - links to the LiveBook documents hosted somewhere else
- `wiki` - links to Wikipedia or internal Wikis without typing whole URL

I think that some people would even come with creative ways to abuse that for weird purposes, but in general I find it pretty safe and backwards compatible way to support wide range of potential links.

**Comment by wojtekmach** [comment]
Extensibility via custom schemas is a really nice idea. Similar to previous idea, I don't think it looks good when rendered by GitHub though:

<img width="302" alt="image" src="https://github.com/elixir-lang/ex_doc/assets/76071/d885629c-57b0-40f8-8887-3b8accef6980">

<img width="236" alt="image" src="https://github.com/elixir-lang/ex_doc/assets/76071/b0eea2aa-cb43-4584-b387-a54ba1376e64">

and I personally consider rendering markdown by non-ExDoc tools like GitHub to be crucial.

> So instead of introducing some new magic syntax we will just extend existing one.

I'd push back on "new magic syntax" a bit. I'd say it's "implicit". With what I'm proposing, let's say when generating docs for https://hexdocs.pm/elixir,

```markdown
This is a [`String`].
```

is just a _shortcut_ for:

```markdown
This is a [`String`].

[`String`]: https://hexdocs.pm/elixir/String.html
```

And in the absence of this `` [`String`]: ...``, GitHub would render it as:

<img width="229" alt="image" src="https://github.com/elixir-lang/ex_doc/assets/76071/a613902b-348c-483f-a532-c129f6697f3b">

and I'd say it's pretty decent but it's subjective.

This is an aside, but to make the point the syntax I propose is pretty natural, I have this in Req README.md:

```markdown
(...)

  * An easy to use high-level API: [`Req.request/1`], [`Req.new/1`], [`Req.get!/2`], [`Req.post!/2`], etc.

(...)

[`Req.request/1`]: https://hexdocs.pm/req/Req.html#request/1
```

and this way my README renders nicely on both ExDoc and GitHub:

![image](https://github.com/elixir-lang/ex_doc/assets/76071/d466454b-1a69-4968-8107-240c3c3db651)

![image](https://github.com/elixir-lang/ex_doc/assets/76071/a7fbd436-c0d9-44aa-9898-5bf3088df0fb)

and if I was able to change the markdown to something like this:

```markdown
(...)

  * An easy to use high-level API: [`Req.request/1`], [`Req.new/1`], [`Req.get!/2`], [`Req.post!/2`], etc.

(...)

<!-- ex_doc:begin-ignore -->

[`Req.request/1`]: https://hexdocs.pm/req/Req.html#request/1

<!-- ex_doc:end-ignore -->
```

then I could have cake and eat it too. ExDoc would resolve these links by itself, e.g. `` [`Req.request/1`] `` would point to `Req.html#request/1`, a _relative_ link, which means package versioning will just work. And then non-ExDoc will resolve links too.

**Comment by hauleth** [comment]
Well, my proposal is to leave "regular" syntax of `` `String` `` as is, just add support for some extra stuff in that rare cases when needed. Your proposal is nice, but that has a problem that it uses CommonMark parsing problems. I think that many such "handling" references can cause rendering slowdowns as it will require many additional lookups for the defined links. While `<xref:String>` do not render nicely I think that in many cases it will not be used, it will be more of an escape hatch. In the same way with my proposal you can do:

```
[`Req.request/1`](xref:Req.request/1)
```

If you worry about rendering in the GitHub or other non-ExDoc rendering tool. It will create dangling link, but will render correctly like that:

[`Req.request/1`](xref:Req.request/1)

<img width="158" alt="image" src="https://github.com/elixir-lang/ex_doc/assets/291639/67d4e855-7470-4a64-97df-161051b34cc3">

<sub>It seems that GitHub strips unknown schema reference as well</sub>

**Comment by wojtekmach** [comment]
If it's mostly an escape hatch, what would be some examples for xref: and elixir: links?

It's up in the air still, but if Erlang adopts markdown and this proposal, we could refer to the file module as `` [`file`] `` (or `` [`:file`] `` from Elixir-style docs) as opposed to, say, `` [`file`](xref:file) `` and I'd consider that a win.

**Comment by hauleth** [comment]
`xref` is simple:

You have document like `a.md` and you want to link to section `#foo` in document `b.md` or options list `#options` in module `Foo.Bar`. Currently there is no way AFAIK to do such cross-document references, especially when you want to rename them via `[]()` syntax.

Same for `elixir`:

Imagine that you [are writing library that redefines `|>`](https://hexdocs.pm/magritte/Magritte.html#%7C%3E/2) and you want to link to Elixir documentation for some parts of the document. Or you want to link for example to [compatibility guide](https://hexdocs.pm/elixir/compatibility-and-deprecations.html) from your documentation. Obviously you can use fully qualified link, but good shorthand can became quite useful.

**Comment by wojtekmach** [comment]
`` [See "foo" section in b docs](b.md#foo) `` totally works. We even automatically change .md to .html for html and .xhtml for epub. :)

For custom |>, when referring to the OG one, I'd explicitly write `` `Kernel.|>/2` `` (or `` [`Kernel.|>/2`] ``), that is.)

> Or you want to link for example to [compatibility guide](https://hexdocs.pm/elixir/compatibility-and-deprecations.html) from your documentation

Yeah, good point!

**Comment by garazdawi** [comment]
I agree that [`file`] would be a good addition for Erlang markdown docs.

When working with the Markdown convention for Erlang/OTP I imagined that you could write links like this:

> ## Links
> 
> When writing documentation in markdown links are automatically found in any
> inline code segment that looks like an MFA. For example:
> 
> ```
> -doc "See `sub/2` for more details"
> ```
> 
> will create a link to the `sub/2` function in the current module if it exists.
> You can also use `` `sub/2` `` as the link target. For example:
> 
> ```
> -doc "See [subtract](`sub/2`) for more details"
> -doc "See [`sub/2`] for more details"
> -doc "See [subtract] for more details
> 
> [substract]: `sub/2`
> "
> -doc "See [subtract][1] for more details
> 
> [1]: `sub/2`
> "
> ```
> 
> The above will all result in the same link being created.

https://garazdawi.github.io/otp/system/documentation.html#links

I think that is the same as what you have proposed.

**Comment by josevalim** [comment]
To be honest, I like `m:math` in your proposal more than ``[`math`]`` (please don't murder me @wojtekmach).

The other thing is that you link pages today using `../path/to/the/page.md`. This at least means the link also works on GitHub but you can't have duplicate page names (because we deploy everything into a flat namespace).

**Comment by wojtekmach** [comment]
@garazdawi to clarify, my proposal is for:

```erl
-doc "See `sub/2` for more details"
```

to NOT create a link. The link needs to be created with:

```erl
-doc "See [`sub/2`] for more details"
```

I agree with the rest of your message 100%.

It is slightly more verbose but explicit and that would be a good thing. If, say, you write `` [`foo`] `` but the module does not exist we warn/error. And if you don't want to link to a module that doesn't exist, don't link, i.e. write `` `foo` `` instead. So we no longer need to guess what the user wanted to write.

> The other thing is that you link pages today using ../path/to/the/page.md.

@josevalim right. I think we can solve it by normalising the path off the project current directory, e.g.: `../path/to/the/page.md` becomes `path-to-the-page.html`

**Comment by wojtekmach** [comment]
`m:module` is explicit too so that's fine by me. As mentioned earlier, using the `` [ref] `` notation gives us ability to write .md files that have links that work when rendered by ExDoc AND GitHub. It is slightly annoying because one needs to write these links explicitly (in order for them to work on GitHub), e.g. `some.md`:


```markdown
See [file] module.


[file]: https://www.erlang.org/doc/man/file
```

I wouldn't use this workaround every single time but I think it is nice to have such option.

**Comment by josevalim** [comment]
Yeah, I would say `m:foo` is also pretty unambiguous, easier to type, and works with everything else we have today, so I am happy to close this one.

**Comment by garazdawi** [comment]
I'm fine with only adding `m:foo` i ex_doc. The only reason I liked the [`foo`] approach was because Rust uses it and when it is possible I prefer to not invent something new when someone else already has started a standard of sorts.

> The other thing is that you link pages today using `../path/to/the/page.md`. This at least means the link also works on GitHub but you can't have duplicate page names (because we deploy everything into a flat namespace).

Getting links to work on Github and in the docs for cross-application references is not very easy. Normally the doc source is located in "app/doc" or even "app/doc/src". So if we are to do a relative link that works in github for that we have to do `../../app/doc/page.md`, while in ex_doc the link would be `../app/page.md`. For our "erts" and "system" applications, things gets even more weird.

By adding `p:app:page.md` we can use the already existing `:deps` configuration to tell ex_doc what the path to an application should be, so that we don't have to know that when we are writing the docs. These links however only need to be in the link target, so maybe there is a better way to do it?

**Comment by josevalim** [comment]
Ah, good call on linking across apps. That's not possible today. The only unfortunate caveat is that we call these additional pages in ExDoc "extras". That's because different packages give different names to them, so maybe it should be `e:app:page.md`, but I assume that's minor.

**Comment by garazdawi** [comment]
Maybe it doesn't even have to have a prefix, but could just be `app:page.md`? The .md extension would signal that it is an extra/page. Anyway, I'm fine with using `e:` as well.

Unfortunately it seems like you are not allowed by gfm to have semi-colon as part of a reference, so it is not possible to create links that work both in Github and ExDoc using this syntax. That is the below does not create a link:

```
See [app and page](e:app:page.md)

[e:app:page.md]: ../../app/page.md
```

Seems like any symbols except `:` is allowed... so we could use `/` or `.` if we think creating these types of links are important. I doubt that we will use it for Erlang/OTP docs, but I can see the usefulness in any umbrella style project.

I suppose the same problem applies to type/callback links and MFA links in Erlang docs (Elixir works since you use `.` for function calls).

**Comment by josevalim** [comment]
@garazdawi for completeness, we do support ``See [app and page](`foo/2`)`` which means we could support ``See [app and page](`e:app:page.md`)``.

We will also need to decide if we want to include the extension or not on `e:app:page.md`. If we remove the extension, it could be used to link to any page (including a module page). This means we can use the `p` prefix or `l` for link. We can automatically add the relevant extension (either .html or .xhtml for epub).

**Comment by wojtekmach** [comment]
> `See [app and page](e:app:page.md)`

yeah this is unexpected. Would need to introduce and intermediate name, e.g.:

```markdown
See [app and page](app-page.md)

[app-page.md]: ../../app/page.md
```

or simply the link text:

```markdown
See [app and page]

[app and page]: ../../app/page.md
```

but yeah not ideal.

**Comment by wojtekmach** [comment]
oh, by accident I tried the following and it worked:

```markdown
See [app and page][e:app:page.md]

[e:app:page.md]: ../../app/page.md
```

brb need to properly learn markdown I guess!

**Comment by wojtekmach** [comment]
seems the [foo][bar] notation, as opposed to [foo](bar), is for references as opposed to URLs so perhaps it is more appropriate for what we're discussing.

**Comment by wojtekmach** [comment]
OK here's my final stab at it. Please don't murder me @josevalim. :)

## The `` `text` `` notation

Principle: If `` `text` `` looks like a reference we try to create a link for it.

We validate links as much as possible. We warn when we think one made a typo or referenced undocumented things. We have heuristics whether to warn, e.g.: we don't warn when the module does not exist, we warn when the module does exist but is undocumented, when the module exists but the function does _not_, etc. I think catching mistakes is essential to allow people to produce rock solid documentation that Elixir is known for.

Pros:

  * Muscle memory built over the last decade

  * Easy to type

  * Renders as `<code>text</code>` on GitHub

Cons:

  * Ambiguous whether one wanted to create a link or have monospaced font.

  * Say you want to refer to an "Erlang" module. `` `:file` `` does _not_ link because we don't know if one was referring to a module or an atom. At the moment we solve it by writing: `` [`:file`](`:file`) ``.
    `` [`:file`](`:file`) `` renders on GitHub as `<a href="%60:file%60"><code>:file</code></a>` which is terrible.  However, `` [`:file`](m:file) `` renders on GitHub as `<code>:file</code>` which is ideal though the link cannot be later overwritten. But then again one can `` [`:file`] (...) [`:file`]: ... `` or ``[see `:file`][whatevs] ... [whatevs]: ...`` and it just works.

  * Say you wrote in your changelog:

        Deprecate `Map.size/1` in favour of `Kernel.map_size/1`.

    ExDoc warns because you are referring to a deprecated function. This warning can be suppressed by setting `skip_undefined_reference_warnings_on` or `skip_code_autolink_to`.

  * Say you wrote:

        Unfortunately we cannot have `MapSet.new/2` because (...)

    ExDoc warns because `MapSet` module exists but it does not have function `new/2`.

    Similarly you wrote:

        (...) Ruby's `Set` (...)

    ExDoc warns because `Set` module is deprecated.

  * Say you make a typo and write `` `Genserver` ``. ExDoc sees the module does not exist so it does _not_ warn. We could probably catch this though. But maybe we shouldn't.  Maybe sometimes we want to warn (we made a typo) and sometimes we don't want, it's not a typo but making a point.

## The `` [`text`] `` notation

Principle: If `` [`text`] `` does not have a corresponding `` [`text`]: link``, we assume `text` is a reference and if it's valid we link and otherwise we warn.

Similarly tu rustdoc, the notation without backtics is also supported with the same rules: `` [Atom] `` would render as `<a href=...>Atom</a>`. This is useful because sometimes we may not necessarily want to use the monospace font but stil link. For example, `[app:phoenix_live_view]` could be rendered as `<a href="https://hexdocs.pm/phoenix_live_view">Phoenix LiveView</a>`.

Pros:

  * Markdown uses `` [text][ref] `` and `` [ref] (...) [ref]: url`` notation for references. We're just dropping the "right-hand side", inferring it.

  * We never accidentally attempt to link and thus produce wrong warnings. Our warnings are 100% justified.

  * Less code complexity, no need to have heuristics and can have simpler and stricter reference string parser.

  * Can remove `skip_undefined_reference_warnings_on`

  * Can remove `skip_code_autolink_to`

  * Rust does it. Maybe/probably others will follow. Btw, Obsidian uses `[[title]]`.

Cons:

  * Muscle memory built over the last decade

  * All existing docs over the last decade used `` `text` `` notation and ideally, at some point, would have to be updated.

  * Slightly more verbose to type, need to use brackets.

  * Not standardized. Need to patch EarmarkParser (done!), cmark if one wants to use that, etc.

  * If the user wanted to write literally `` [foo] `` we'd consider that a reference and link when not anticipated or more likely warn.
    Need to see how Rust handles it but my guess is that it is not a big deal in practice.

  * `` [`:file`] `` renders as `[<code>file</code>]` (with brackets) on GitHub. It is much better than the `` [`:file`](`:file`) `` mentioned earlier though.

## Transition

If we move forward with this, ExDoc would definitely keep old behaviour alongside until v1.0. I think we can create tools that help with the transition, for example if `` `x` `` would have created a link, it gets reritten to `` [`x`] ``, the tool is written in a way it can run on CI, etc.

## Bonus

As I alluded to earlier, I think it's really nice to have links on both ExDoc and GitHub. Here's an example:

  * https://hexdocs.pm/req/readme.html
  * https://github.com/wojtekmach/req/blob/main/README.md
  * https://hexdocs.pm/req/changelog.html
  * https://github.com/wojtekmach/req/blob/main/CHANGELOG.md

I do this by manually inserting references like this:

```markdown
... [`Req.new/1`] ...

[`Req.new/1`]: https://hexdocs.pm/req/Req.html#new/1
```

This is somewhat annoying but actually less hassle than one might think. I only mention this because in order for this to work, we MUST use the `[ref]` notation. So it's a matter of consistency and even allows doing the `@moduledoc File.read!(...)` trick. I can easily imagine a tool that could even automatically add these "footer refs" to markdown documents, to keep links in sync so to speak.

## Conclusion

If we could do it all over again, it is subjective but I believe ``[`:file`]`` would have been a pretty clear winner. I actually think the notation is intuitive. The implementation would definitely be simpler and we could catch more mistakes. We can't do it all over again, but I think we can build tools that will help moving us forward.

**Comment by wojtekmach** [comment]
Here's another idea. Say we have:

```elixir
@doc """
See [`:file`].
"""
```

When Elixir creates the docs chunk, it could add to the chunk the equivalent of something along the lines of:

```elixir
@doc refs: %{
  "`:file`" => {:module, :kernel, :file}
}
```

and _Elixir_ could validate these links, emit diagnostics that IDEs can pick up. Then in ExDoc it's trivial to render these, we know they're good so we literally take the existing markdown piece and append, you guessed it,

```markdown
[`:file`]: https://www.erlang.org/doc/man/file
```

and it just works. I believe we want to do this part of the work, building the final URL, in ExDoc because we may want to change base url on occasion.

The only thing is we want to also handle links in _extras_ so if the link collection and validation logic is in Elixir, we'd need to expose APIs so ExDoc can use them. This might be a deal breaker.

Anyway, if rendering is trivial in ExDoc, then it's trivial in Livebook, LSPs, etc too.

It's obviously a very high bar to entry to end up in the Elixir itself, but something to maybe ponder. Could of course start as ExDoc's Mix.Tasks.Compile.Docs or something like that.

**Comment by whatyouhide** [comment]
> Similarly tu rustdoc, the notation without backtics is also supported with the same rules: [Atom] would render as `<a href=...>Atom</a>`.

I don't see the point of this. If you want the non-monospaced font and still link, maybe you should do `[Atom](m:Atom)`?

> Not standardized. Need to patch EarmarkParser (done!), cmark if one wants to use that, etc.

`` `Atom` `` linking to the `Atom` module is also not standardized. `[foo]` is in the Markdown spec, it'll just try to find a `[foo]: ...` reference somewhere in the docs. Right?

---

Apologies if I’m saying something dumb, but where did the `[prefix:...]` notation go? I really liked the idea of having

```
[m:Atom] → links to the Atom module
```

and friends. @josevalim and @wojtekmach: please don't murder me, either of you 🔪 😄

**Comment by josevalim** [comment]
Folks, let's just please drop this. If we had a time machine, _sure_, but the reality of breaking all Elixir documentation written so far is just not acceptable. We are being very careful with not having any breaking changes to the language and then we decide to make _all docs stale_?

The truth is that we would most likely never be able to fully remove the `` `foo/2` `` notation, so we would effectively end-up with two syntaxes, not effectively mitigating their cons, and not fully leveraging the pros of both.

The whole point of `foo/2` was precisely so simply referring to the code was enough to make it become a link because _that is going to be more useful to the reader_. The documentation writer may need to deal with some false positives, but making links slightly harder to write has the potential of having fewer links and therefore a worse user experience.

In any case, we should still introduce `m:file` for Erlang and something else to link to pages. I will open up a separate issue. And those who want to use the ``[`Req.new/1`]: https://hexdocs.pm/req/Req.html#new/1`` notation can do it today anyway.

> Say you make a typo and write `Genserver`. ExDoc sees the module does not exist so it does not warn. We could probably catch this though. But maybe we shouldn't. 

This is just a bug. It should be addressed.

> When Elixir creates the docs chunk, it could add to the chunk the equivalent of something along the lines of:

Making the Elixir compiler parse markdown will just slow it down.

**Comment by whatyouhide** [comment]
> Folks who prefer explicitness can use m:GenServer, which will catch bugs

Ah this part I did not get! Awesome, that's all I was talking about 🙃 Thanks José.

**Comment by wojtekmach** [comment]
@josevalim please don't kill me. :D random: I was just listening to a podcast and one person said something along the lines of we're all boiling frogs and I thought about this issue for some reason.

To take a step back, for sure we need a `m:<module>` to conveniently link to "Erlang". `m:` can help catching docs bugs too.

I think catching bugs is really helpful.

However, unless we introduce some other modifier, we won't be able to catch bugs like:

- `` `a_local_functio/0` `` - should be `` `a_local_function/0` ``.
- `` `binarypart/3` `` - should be `` `binary_part/3` ``
- `` `GenServe.call/3` `` - should be `` `GenServer.call/3` ``
- repeat for types and callbacks.


I'd like to propose the following:

- we keep `` `ref` `` as is.
- we introduce a `` [`ref`] `` which is equivalent to `` [`ref`](`ref`) ``. It is two extra characters but we are more intentional, we catch all bugs, it's familiar to Rust people, and it renders OK on GitHub.

We could even add an option, off by default, that turns off auto-linking `` `ref` `` so people don't have to use `skip_undefined_reference_warnings_on` or `skip_code_autolink_to`. This option could even be on by default for "Erlang" projects since there is less existing docs.

The downside is we have two syntaxes to do the same thing but it's not quite the same thing, isn't it, they do different things. If, and that's a big if, we ever have a tool that rewrites `` `ref` `` to `` [`ref`] ``, a potential concern is that would generate a ton of code churn. But we survived introduction of `mix format` (and some projects never adopted which is totally fine) so we'd survive that for sure too.

wdyt?
